---
title: "Single Nucleus Sequencing Trial Quality Control and Analysis"
output:
  pdf_document: default
  html_document: default
---
--
```{r include = FALSE}
### Load data and packages
require(Matrix)
library(Seurat)
require(myUtils)
require(Biomart)
require(ggplot2)
require(Hmisc)

# Load snSeq data 1:
data1 = readMM('data/snSeqQC/cellranger302_count_29507_5705STDY7945423_mm10-3_0_0_premrna/filtered_feature_bc_matrix/matrix.mtx.gz')
rowdata1 = read.delim('data/snSeqQC/cellranger302_count_29507_5705STDY7945423_mm10-3_0_0_premrna/filtered_feature_bc_matrix/features.tsv.gz', header = FALSE)
coldata1 = read.delim('data/snSeqQC/cellranger302_count_29507_5705STDY7945423_mm10-3_0_0_premrna/filtered_feature_bc_matrix/barcodes.tsv.gz', header = FALSE)
data1 = as.matrix(data1)
rownames(data1) = rowdata1[,2]
colnames(data1) = coldata1[,1]

# Load snSeq data 2:
data2 = readMM('data/snSeqQC/matrix.mtx.gz')
rowdata2 = read.delim('data/snSeqQC/features.tsv.gz', header = FALSE)
coldata2 = read.delim('data/snSeqQC/barcodes.tsv.gz', header = FALSE)
data2 = as.matrix(data2)
rownames(data2) = rowdata2[,2]
colnames(data2) = coldata2[,1]
colnames(data2) = paste(colnames(data2),'2', sep = '')

# Load Allen Data:
options(stringsAsFactors = FALSE)
dataAllen = read.delim('data/Allen/mouse_VISp_2018-06-14_exon+intron_cpm-matrix.csv', sep = ',')
rowdataA = read.delim('data/Allen/mouse_VISp_2018-06-14_genes-rows.csv', sep = ',')
coldataA = read.delim('data/Allen/mouse_VISp_2018-06-14_samples-columns.csv', sep = ',')
rownames(dataAllen) = rowdataA[,1]
commonGenes = intersect(intersect(rownames(dataAllen), rownames(data1)), rownames(data2))
celltypes = coldataA[,'class']
keep = celltypes %in% c('GABAergic', 'Endothelial', 'Glutamatergic', 'Non-Neuronal')
coldataA = coldataA[keep,]
dataAllen = dataAllen[,keep]
celltypes = celltypes[keep]
subtypes = unlist(lapply(1:dim(coldataA)[1], function(x) paste(coldataA[x,c('class', 'subclass')], sep = '_', collapse = '_')))

data1 = data1[commonGenes,]
data2 = data2[commonGenes,]
dataAllen = dataAllen[commonGenes,]
dataAllen = dataAllen[,2:dim(dataAllen)[2]]
```

The starting point for the analysis is the filtered output form the cellranger software. The software assigns all counts with the same barcode to one cell. In addition, it discards reads from barcodes that occur only a few number of times, as these likely correspond to read errors and not cells.

## Quality Control

I did not remove any more cells based on low numbers of detected genes. I also did not remove cells as potential doublets based on a high numbers of detected genes alone.

```{r, include = TRUE}
# Quality Control Statistics:
genesDetected1 = colSums(data1 > 0)
genesDetected2 = colSums(data2 > 0)
readsDetected1 = colSums(data1)
readsDetected2 = colSums(data2)
numberOfCells1 = length(genesDetected1)
numberOfCells2 = length(genesDetected2)
qcDataFrame = data.frame(genesDetected = c(genesDetected1, genesDetected2),
                         readsDetected = c(readsDetected1, readsDetected2),
                         Run = c(rep('FFT4G', length(genesDetected1)), rep('OCT1', length(genesDetected2))))
p1 <- ggplot(qcDataFrame, aes(x=Run, y=genesDetected, fill = Run)) + 
  geom_violin() + 
geom_jitter(shape=16, size = 0.25, width = 0.45, height = 0.45) +
  stat_summary(fun.data=mean_sdl, geom="pointrange", color="red") +
  annotate("text", x = 0.7, y = 5000, label = "n = 4865") +
  annotate("text", x = 1.85, y = 5000, label = "n = 5814") + ylab('Genes Detected')
pdf(file = 'NumberOfDetectedGenes.pdf', width = 10, height = 10)
p1
dev.off()
p1
```